README.md
(Fetched from https://pineight.com/)

IF THIS FILE HAS NO LINE BREAKS:  View it in a web browser.
Windows Notepad is a very popular text editor that comes with the
Windows operating system, but it doesn't recognize line breaks in
text files made on Linux or any other UNIX-like operating system.
Text files with UNIX line breaks display correctly in WordPad,
Notepad++, Programmer's Notepad, Gedit, or a web browser, just
not in Windows Notepad.


This is a minimal working program for the Nintendo Entertainment
System using the SGROM or SNROM board.

Concepts illustrated:

* init code
* setting up a static background
* structure of a game loop
* DPCM-safe controller reading
* 8.8 fixed-point arithmetic
* acceleration-based character movement physics
* sprite drawing and animation, with horizontal flipping
* makefile-controlled conversion of sprite sheets
* initializing the MMC1
* loading tile data into CHR RAM
* calls from one PRG bank to another

Setting up the build environment
================================

Building this demo requires that the following software be
installed on your computer first:

* ca65 and ld65, the assembly language tools that ship
  with the cc65 C compiler
* Python, a programming language interpreter
* Python Imaging Library, a Python extension to read and write
  bitmap images
* GNU Make, a program to calculate which files need to be
  rebuilt when other files change
* GNU Coreutils, a set of simple command-line utilities for
  file management and text processing

It also requires general familiarity with the command prompt.
You are encouraged to read and understand the articles on general
computer science topics listed at [Before the basics].

[Before the basics]: http://wiki.nesdev.com/w/index.php/Before_the_basics


On Windows
----------

The MSYS project ports Make, Coreutils, and other key parts of
the GNU operating environment to Windows.  An easy way to install
MSYS is through the automated installer provided by devkitPro.

1. Visit [MSYS2 Project home page].
2. Follow the instructions there to download and run the
   MSYS2 Installer.

To install Python under Windows:

1. Visit [Python home page].
2. Under Downloads, under Download for Windows, click Python 3.5.1.
   (This step may change if the maintainers of the Python web site
    reorganize the Python web site.)
3. In your web browser's downloads folder, run `python-3.5.1.msi`
4. Follow the prompts through the installer wizard.

To install Python Imaging Library under Windows:

1. Visit [Pillow project page].
2. Under Package, click the latest version.
3. Click Downloads.
4. Click `Pillow-3.2.0.win32-py3.5.exe`. The `py-3.5` part MUST
   match the major and minor version of Python installed on your PC.
   (In Python 3.5.1, the major version is 3, the minor version is 5,
   and the bug fix version is 1.)
5. In your web browser's downloads folder, run
   `Pillow-3.2.0.win32-py3.5.exe`
6. Follow the prompts through the installer wizard

To install cc65 under Windows:

1. Visit [cc65 introduction].
2. Scroll to the bottom.
3. Click "Windows Snapshot" to download a zip file.
4. Open the zip file.
5. Inside the zip file, open the bin folder.
6. Drag `ca65.exe` and `ld65.exe` into a new folder.
7. Add this folder to your "Path" environment variable.

Finally, to make ca65, ld65, and Python available to Make, you'll
need to add the folders containing `ca65.exe`, `ld65.exe`, and
`python.exe` to the `Path` environment variable.  Because the steps
for setting environment variables differ between versions of Windows,
you'll want to search the web for `windows x.x path variable`,
replacing `x.x` with `7`, `8.1`, etc.

Finally, open the makefile in a text editor and change EMU to
whatever NES emulator you have installed.

[MSYS2 Project home page]: https://msys2.github.io/
[Python home page]: https://www.python.org/
[Pillow project page]: https://pypi.python.org/pypi/Pillow
[cc65 introduction]: http://cc65.github.io/cc65/

Organization of the program
===========================

Include files
-------------

* `nes.inc`: Register definitions and useful macros
* `global.inc`: Global variable and function declarations

Source code files
-----------------

* `mmc1.s`: iNES header and driver for MMC1
* `unrom.s`: iNES header and driver for UNROM/UOROM
* `init.s`: PPU and CPU I/O initialization code
* `main.s`: Main program
* `bankcalltable.s`: List of entry points called through a far call
  (one that goes from one bank to another)
* `chrram.s`: CHR RAM data setup
* `bg.s`: Background graphics setup
* `player.s`: Player sprite graphics setup and movement
* `pads.s`: Read the controllers in a DPCM-safe manner
* `ppuclear.s`: Useful subroutines for interacting with the S-PPU

Each source code file is made up of subroutines that start with
`.proc` and end with `.endproc`.  See the [ca65 Users Guide] for
what these mean.

[ca65 Users Guide]: http://cc65.github.io/doc/ca65.html

The tools
=========

The `tools` folder contains a couple essential command-line programs
written in Python to convert graphics into a form usable by the NES.
The makefile contains instructions to run the program again whenever
the original asset data changes.

* `pilbmp2nes.py` is a program to convert bitmap images in PNG or
  BMP format into tile data usable by several classic video game
  consoles.  It has several options to control the data format; use
  `pilbmp2nes.py --help` from the command prompt to see them all.

Legal
=====

The demo is distributed under the following license, based on the
GNU All-Permissive License:

> Copyright 2011-2015 Damian Yerrick
>
> Copying and distribution of this file, with or without
> modification, are permitted in any medium without royalty provided
> the copyright notice and this notice are preserved in all source
> code copies.  This file is offered as-is, without any warranty.



CHANGES.txt
0.06 (2016-06-24)
* python 2.x to 3.x instructions
* make clean clean obj also, silently + explicitly
* make clean clean generated map.txt, mapalt.txt
* shuffled source and object dirs a little
* introduced target directory for output
* removed zip target

0.05 (2015-02-13)
* renamed .h files to .inc so as not to confuse with C headers
* broke parts of main.s into init.s, bg.s, and player.s
* added overall map of source code to README.md
* moved init code into init.s
* indented forward branches in player.s

0.04 (2014-11-09)
* expanded build instructions (requested by whicker)
* switched to 7-frame walk

0.03 (2014-05-17)
* corrected character's head position
* corrected negative byte values for ca65 2.14
* moved -o to front of linker command line for ld65 2.14

0.02 (2011-07-23)
* added comment about the holy war over clearing or not clearing RAM
* added notice about line breaks in Windows Notepad
  (requested by 3GenGames)
* added SNROM version with CHR RAM and inter-bank calls

0.01 (2011-07-20)
* initial release


source/bg.s
.include "nes.inc"
.include "global.inc"

.proc draw_bg
  ; Start by clearing the first nametable
  ldx #$20
  lda #$00
  ldy #$AA
  jsr ppu_clear_nt

  ; Draw a floor
  lda #$23
  sta PPUADDR
  lda #$00
  sta PPUADDR
  lda #$0B
  ldx #32
floorloop1:
  sta PPUDATA
  dex
  bne floorloop1
  
  ; Draw areas buried under the floor as solid color
  ; (I learned this style from "Pinobee" for GBA.  We drink Ritalin.)
  lda #$01
  ldx #5*32
floorloop2:
  sta PPUDATA
  dex
  bne floorloop2

  ; Draw blocks on the sides, in vertical columns
  lda #VBLANK_NMI|VRAM_DOWN
  sta PPUCTRL
  
  ; At position (2, 20) (VRAM $2282) and (28, 20) (VRAM $229C),
  ; draw two columns of two blocks each, each block being 4 tiles:
  ; 0C 0D
  ; 0E 0F
  ldx #2

colloop:
  lda #$22
  sta PPUADDR
  txa
  ora #$80
  sta PPUADDR

  ; Draw $0C $0E $0C $0E or $0D $0F $0D $0F depending on column
  and #$01
  ora #$0C
  ldy #4
tileloop:
  sta PPUDATA
  eor #$02
  dey
  bne tileloop

  ; Columns 2, 3, 28, and 29 only  
  inx
  cpx #4  ; Skip columns 4 through 27
  bne not4
  ldx #28
not4:
  cpx #30
  bcc colloop

  ; The attribute table elements corresponding to these stacks are
  ; (0, 5) (VRAM $23E8) and (7, 5) (VRAM $23EF).  Set them to 0.
  ldx #$23
  lda #$E8
  ldy #$00
  stx PPUADDR
  sta PPUADDR
  sty PPUDATA
  lda #$EF
  stx PPUADDR
  sta PPUADDR
  sty PPUDATA

  rts
.endproc



source/chrram.s
;
; Trivial CHR RAM loader for NES
; Copyright 2011 Damian Yerrick
;
; Copying and distribution of this file, with or without
; modification, are permitted in any medium without royalty provided
; the copyright notice and this notice are preserved in all source
; code copies.  This file is offered as-is, without any warranty.
;
.include "nes.inc"
.include "mmc1.inc"
.export load_chr_ram_far

.segment "BANK13"
template_chr:
  .incbin "object/nes/bggfx.chr"
  .incbin "object/nes/spritegfx.chr"

;;
; Loads 8192 bytes of uncompressed data into CHR RAM.
.proc load_chr_ram_far
srclo = 0
srchi = 1

  lda #<template_chr
  sta srclo
  lda #>template_chr
  sta srchi
  ldy #0
  sty PPUADDR  ; set starting location in CHR RAM to $0000
  sty PPUADDR
  ldx #>8192  ; X counts the remaining length of CHR in 256-byte units
loop:
  lda (srclo),y
  sta PPUDATA
  iny
  bne loop
  ; after every 256th byte we end up here
  inc srchi  ; move on to the next set of 256 bytes of CHR
  dex
  bne loop
  jmp bankrts
.endproc




source/player.s
.include "nes.inc"
.include "mmc1.inc"
.include "global.inc"

.segment "ZEROPAGE"
; Game variables
player_xlo:       .res 1  ; horizontal position is xhi + xlo/256 px
player_xhi:       .res 1
player_dxlo:      .res 1  ; speed in pixels per 256 s
player_yhi:       .res 1
player_facing:    .res 1
player_frame:     .res 1
player_frame_sub: .res 1

; constants used by move_player
; PAL frames are about 20% longer than NTSC frames.  So if you make
; dual NTSC and PAL versions, or you auto-adapt to the TV system,
; you'll want PAL velocity values to be 1.2 times the corresponding
; NTSC values, and PAL accelerations should be 1.44 times NTSC.
WALK_SPD = 105  ; speed limit in 1/256 px/frame
WALK_ACCEL = 4  ; movement acceleration in 1/256 px/frame^2
WALK_BRAKE = 8  ; stopping acceleration in 1/256 px/frame^2

LEFT_WALL = 32
RIGHT_WALL = 224

.segment "CODE"

.proc init_player
  lda #0
  sta player_xlo
  sta player_dxlo
  sta player_facing
  sta player_frame
  lda #48
  sta player_xhi
  lda #192
  sta player_yhi
  rts
.endproc

;;
; Moves the player character in response to controller 1.
.proc move_player

  ; Acceleration to right: Do it only if the player is holding right
  ; on the Control Pad and has a nonnegative velocity.
  lda cur_keys
  and #KEY_RIGHT
  beq notRight
  lda player_dxlo
  bmi notRight
  
    ; Right is pressed.  Add to velocity, but don't allow velocity
    ; to be greater than the maximum.
    clc
    adc #WALK_ACCEL
    cmp #WALK_SPD
    bcc :+
      lda #WALK_SPD
    :
    sta player_dxlo
    lda player_facing  ; Set the facing direction to not flipped 
    and #<~$40         ; turn off bit 6, leave all others on
    sta player_facing
    jmp doneRight
  notRight:

    ; Right is not pressed.  Brake if headed right.
    lda player_dxlo
    bmi doneRight
    cmp #WALK_BRAKE
    bcs notRightStop
    lda #WALK_BRAKE+1  ; add 1 to compensate for the carry being clear
  notRightStop:
    sbc #WALK_BRAKE
    sta player_dxlo
  doneRight:

  ; Acceleration to left: Do it only if the player is holding left
  ; on the Control Pad and has a nonpositive velocity.
  lda cur_keys
  and #KEY_LEFT
  beq notLeft
  lda player_dxlo
  beq isLeft
    bpl notLeft
  isLeft:

    ; Left is pressed.  Add to velocity.
    lda player_dxlo
    sec
    sbc #WALK_ACCEL
    cmp #256-WALK_SPD
    bcs :+
      lda #256-WALK_SPD
    :
    sta player_dxlo
    lda player_facing  ; Set the facing direction to flipped
    ora #$40
    sta player_facing
    jmp doneLeft

    ; Left is not pressed.  Brake if headed left.
  notLeft:
    lda player_dxlo
    bpl doneLeft
    cmp #256-WALK_BRAKE
    bcc notLeftStop
    lda #256-WALK_BRAKE
  notLeftStop:
    adc #8-1
    sta player_dxlo
  doneLeft:

  ; In a real game, you'd respond to A, B, Up, Down, etc. here.

  ; Move the player by adding the velocity to the 16-bit X position.
  lda player_dxlo
  bpl player_dxlo_pos
    ; if velocity is negative, subtract 1 from high byte to sign extend
    dec player_xhi
  player_dxlo_pos:
  clc
  adc player_xlo
  sta player_xlo
  lda #0          ; add high byte
  adc player_xhi
  sta player_xhi

  ; Test for collision with side walls
  cmp #LEFT_WALL-4
  bcs notHitLeft
    lda #LEFT_WALL-4
    sta player_xhi
    lda #0
    sta player_dxlo
    beq doneWallCollision
  notHitLeft:

  cmp #RIGHT_WALL-12
  bcc notHitRight
    lda #RIGHT_WALL-13
    sta player_xhi
    lda #0
    sta player_dxlo
  notHitRight:

  ; Additional checks for collision, if needed, would go here.
doneWallCollision:

  ; Animate the player
  ; If stopped, freeze the animation on frame 0
  lda player_dxlo
  bne notStop1
    lda #$C0
    sta player_frame_sub
    lda #0
    beq have_player_frame
  notStop1:

  ; Take absolute value of velocity (negate it if it's negative)
  bpl player_animate_noneg
    eor #$FF
    clc
    adc #1
  player_animate_noneg:

  lsr a  ; Multiply abs(velocity) by 5/16
  lsr a
  sta 0
  lsr a
  lsr a
  adc 0

  ; And 16-bit add it to player_frame, mod $600  
  adc player_frame_sub
  sta player_frame_sub
  lda player_frame
  adc #0  ; add only the carry

  ; Wrap from $800 (after last frame of walk cycle)
  ; to $100 (first frame of walk cycle)
  cmp #8  ; frame 0: still; 1-7: scooting
  bcc have_player_frame
    lda #1
  have_player_frame:

  sta player_frame
  rts
.endproc


; We put this in bank 2 to test the bankcall mechanism
.segment "BANK02"
.export draw_player_sprite_far
;;
; Draws the player's character to the display list as six sprites.
; In the template, we don't need to handle half-offscreen actors,
; but a scrolling game will need to "clip" sprites (skip drawing the
; parts that are offscreen).
.proc draw_player_sprite_far
draw_y = 0
cur_tile = 1
x_add = 2         ; +8 when not flipped; -8 when flipped
draw_x = 3
rows_left = 4
row_first_tile = 5
draw_x_left = 7

  lda #3
  sta rows_left
  
  ; In platform games, the Y position is often understood as the
  ; bottom of a character because that makes certain things related
  ; to platform collision easier to reason about.  Here, the
  ; character is 24 pixels tall, and player_yhi is the bottom.
  ; On the NES, sprites are drawn one scanline lower than the Y
  ; coordinate in the OAM entry (e.g. the top row of pixels of a
  ; sprite with Y=8 is on scanline 9).  But in a platformer, it's
  ; also common practice to overlap the bottom row of a sprite's
  ; pixels with the top pixel of the background platform that they
  ; walk on to suggest depth in the background.
  lda player_yhi
  sec
  sbc #24
  sta draw_y

  ; set up increment amounts based on flip value
  ; A: actual X coordinate of first sprite
  ; X: distance to move (either 8 or -8)
  lda player_xhi
  ldx #8
  bit player_facing
  bvc not_flipped
  clc
  adc #8
  ldx #(256-8)
not_flipped:
  sta draw_x_left
  stx x_add

  ; the eight frames start at $10, $12, ..., $1E
  ; 0: still; 1-7: scooting
  lda player_frame
  asl a
  ora #$10
  sta row_first_tile

  ; frame 7 is special: the player needs to be drawn 1 unit forward
  ; because of how far he's leaned forward
  lda player_frame
  cmp #7
  bcc not_frame_7
  
  ; here, carry is set, so anything you add will get another 1
  ; added to it, so subtract 1 when constructing the value to add
  ; to the player's X position
  lda #1 - 1  ; facing right: move forward by 1
  bit player_facing
  bvc f7_not_flipped
  lda #<(-1 - 1)  ; facing left: move left by 1
f7_not_flipped:
  adc draw_x_left
  sta draw_x_left
not_frame_7:

  ldx oam_used
rowloop:
  ldy #2              ; Y: remaining width on this row in 8px units
  lda row_first_tile
  sta cur_tile
  lda draw_x_left
  sta draw_x
tileloop:

  ; draw an 8x8 pixel chunk of the character using one entry in the
  ; display list
  lda draw_y
  sta OAM,x
  lda cur_tile
  inc cur_tile
  sta OAM+1,x
  lda player_facing
  sta OAM+2,x
  lda draw_x
  sta OAM+3,x
  clc
  adc x_add
  sta draw_x
  
  ; move to the next entry of the display list
  inx
  inx
  inx
  inx
  dey
  bne tileloop

  ; move to the next row, which is 8 scanlines down and on the next
  ; row of tiles in the pattern table
  lda draw_y
  clc
  adc #8
  sta draw_y
  lda row_first_tile
  clc
  adc #16
  sta row_first_tile
  dec rows_left
  bne rowloop

  stx oam_used
  jmp bankrts
.endproc

